### PR TITLE
Devise token auth

### PR DESCRIPTION
**Pull request by ivanfelfo**
Added devise token auth

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,8 +1,13 @@
...
-    devise_for :users, controllers: {
-      confirmations: 'v1/confirmations'
+    mount_devise_token_auth_for 'User', at: 'users', controllers: {
+      confirmations: 'v1/confirmations',
+      registrations: 'v1/registrations',
+      sessions: 'v1/sessions'
     }
+    # devise_for :users, controllers: {
+    #   confirmations: 'v1/confirmations'
+    # }
```
Comment (martinjaimem) [inline config/routes.rb:10]:
```suggestion
    }
```

File: spec/requests/users/create_spec.rb [hunk 2]
```diff
@@ -16,16 +16,16 @@
...
     it 'will change the user count by 1' do
       expect { subject }.to change(User, :count).by(1)
     end
 
-    it 'sends confirmation email' do
-      expect { subject }.to change(ActionMailer::Base.deliveries, :count).by(1)
-    end
+    # it 'sends confirmation email' do
+    #   expect { subject }.to change(ActionMailer::Base.deliveries, :count).by(1)
+    # end
```
Comment (martinjaimem) [inline spec/requests/users/create_spec.rb:28]:
```suggestion
    it 'sends confirmation email' do
      expect { subject }.to change(ActionMailer::Base.deliveries, :count).by(1)
    end
```

File: db/migrate/20210430182928_add_trackable_fields.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  def change
+    change_table :users, bulk: true do |t|
+      t.integer  :sign_in_count, default: 0, null: false
+      t.datetime :current_sign_in_at
+      t.datetime :last_sign_in_at
+      t.string   :current_sign_in_ip
+      t.string   :last_sign_in_ip
+    end
+  end
+end
```
Comment (martinjaimem) [inline db/migrate/20210430182928_add_trackable_fields.rb:11]: remove this file, trackable is not needed

File: spec/requests/sessions/create_spec.rb [hunk 4]
```diff
@@ -9,15 +9,23 @@
   let(:params) { { user: { email: 'mail@mail.com', password: 'hello123' } } }
 
   context 'when user already exists' do
-    let!(:created_user) { create(:user, email: 'mail@mail.com', password: 'hello123') }
+    let!(:created_user) do
+      create(:user, email: 'mail@mail.com', password: 'hello123',
+                    password_confirmation: 'hello123', confirmed_at: Time.current)
```
Comment (martinjaimem) [inline spec/requests/sessions/create_spec.rb:14]:
```suggestion
                    confirmed_at: Time.current)
```

File: spec/requests/users/create_spec.rb [hunk 5]
```diff
@@ -50,11 +50,14 @@
 
   context 'when not valid' do
     context 'when trying to register a user with an email that already exists' do
-      let!(:created_user) { create(:user, email: 'mail@mail.com') }
+      let!(:created_user) do
+        create(:user, email: 'mail@mail.com', password: 'hello1234',
+                      password_confirmation: 'hello1234')
```
Comment (martinjaimem) [inline spec/requests/users/create_spec.rb:55]:
```suggestion
)
```

File: config/routes.rb (hunk 1 above)
Comment (ivanfelfo) [reply config/routes.rb:10]: 👍

File: spec/requests/users/create_spec.rb (hunk 2 above)
Comment (ivanfelfo) [reply spec/requests/users/create_spec.rb:28]: 👍

File: spec/factories/users_factories.rb [hunk 6]
```diff
@@ -3,6 +3,7 @@
   factory :user do
     email { Faker::Internet.unique.email }
     password { Faker::Internet.password(8) }
-    gender   { rand(0..3) }
+    password_confirmation { :password }
```
[the remark this answers is not in the document]
Comment (ivanfelfo) [reply spec/factories/users_factories.rb:6]: 👍

File: app/models/user.rb [hunk 7]
```diff
@@ -1,14 +1,23 @@
 # frozen_string_literal: true
 
 class User < ApplicationRecord
-  # Include default devise modules. Others available are:
-  # :lockable, :timeoutable, :trackable and :omniauthable
+  # Include default devise modules.
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :validatable, :confirmable
+         :recoverable, :rememberable, :trackable, :validatable, :confirmable
```
[the remark this answers is not in the document]
Comment (ivanfelfo) [reply app/models/user.rb:6]: 👍

File: db/migrate/20210430182928_add_trackable_fields.rb (hunk 3 above)
Comment (ivanfelfo) [reply db/migrate/20210430182928_add_trackable_fields.rb:11]: 👍

File: spec/requests/sessions/create_spec.rb (hunk 4 above)
Comment (ivanfelfo) [reply spec/requests/sessions/create_spec.rb:14]: 👍

File: spec/requests/users/create_spec.rb (hunk 5 above)
Comment (ivanfelfo) [reply spec/requests/users/create_spec.rb:55]: 👍

File: db/migrate/20210415121803_devise_create_users.rb [hunk 8]
```diff
@@ -22,10 +22,10 @@ def change
...
 
       ## Confirmable
-      # t.string   :confirmation_token
-      # t.datetime :confirmed_at
-      # t.datetime :confirmation_sent_at
-      # t.string   :unconfirmed_email # Only if using reconfirmable
+      t.string   :confirmation_token
+      t.datetime :confirmed_at
+      t.datetime :confirmation_sent_at
+      t.string   :unconfirmed_email # Only if using reconfirmable
```
[the remark this answers is not in the document]
Comment (ivanfelfo) [reply db/migrate/20210415121803_devise_create_users.rb:28]: do you mean comment? because if I restore the add_confirmable_fields file, it'll be repeated

**Review by juanmferreira93 — APPROVED** [review]

File: spec/requests/users/create_spec.rb [hunk 9]
```diff
@@ -50,11 +50,13 @@
 
   context 'when not valid' do
     context 'when trying to register a user with an email that already exists' do
-      let!(:created_user) { create(:user, email: 'mail@mail.com') }
+      let!(:created_user) do
+        create(:user, email: 'mail@mail.com', password: 'hello1234')
+      end
 
       it 'responds with 400' do
```
[the remark this answers is not in the document]
Comment (ivanfelfo) [reply spec/requests/users/create_spec.rb:57]: 👍

File: db/migrate/20210415121803_devise_create_users.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (martinjaimem) [reply db/migrate/20210415121803_devise_create_users.rb:28]: comment 👍

**Review by martinjaimem — APPROVED** [review]

**Pull request merged**